src/Database/Driver.php

```diff
@@ -207,6 +216,75 @@ public function getConnection(): PDO
      */
     abstract public function enabled(): bool;
 
+    /**
+     * @inheritDoc
+     */
+    public function execute(string $sql, array $params = [], array $types = []): StatementInterface
+    {
+        $statement = $this->prepare($sql);
+        if (!empty($params)) {
+            $statement->bind($params, $types);
+        }
+        if ($this->logger) {
+            $this->logAndExecute($statement);
+        } else {
+            $statement->execute();
+        }
+
+        return $statement;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function run(Query $query): StatementInterface
+    {
+        $statement = $this->prepare($query);
+        $query->getValueBinder()->attachTo($statement);
+        if ($this->logger) {
+            $this->logAndExecute($statement);
+        } else {
+            $statement->execute();
+        }
+
+        return $statement;
+    }
+
+    /**
+     * Execute the statement and log the query string.
+     *
+     * @param \Cake\Database\StatementInterface $statement Statement to execute.
+     * @return void
+     */
+    protected function logAndExecute(StatementInterface $statement): void
+    {
+        $exception = null;
+        $took = 0.0;
+
+        try {
+            $start = microtime(true);
+            $statement->execute();
+            $took = (float)number_format((microtime(true) - $start) * 1000, 1);
+        } catch (PDOException $e) {
+            $exception = $e;
+        }
+
+        $loggedQuery = new LoggedQuery();
+        $loggedQuery->driver = $this;
+        $loggedQuery->query = $statement->queryString();
+        $loggedQuery->params = $statement->getBoundParams();
+        $loggedQuery->error = $exception;
+        if (!$exception) {
+            $loggedQuery->numRows = $statement->rowCount();
+            $loggedQuery->took = $took;
+        }
+        $this->logger?->debug((string)$loggedQuery, ['query' => $loggedQuery]);
```

Comment:
This logger instance is already checked before called. Should we just call it exclusively so there's only one execute?

Reply:
Yeah I wanted to avoid a method call hence the check is done before call. But static analyzers don't understand that not null check has already been done.

So yes I could remove the check before the method is called or another option is to pass the logger instance as an argument to the method to keep static analyzers happy.

Reply:
I'd rather keep this simple and have a single wrapper for execute + logging until we decide how to reimagine that.

Reply:
Done